### PR TITLE
feat(channels): declarative DriverCapabilities + canMatch foundation (#3186)

### PR DESCRIPTION
**Pull request by zackees**
Foundation slice for #3186 — declarative driver-capability descriptors and the free `fl::canMatch` function. No driver is migrated and no resolver is wired up here; that's the next set of follow-up PRs.

## What lands

**New header `src/fl/channels/capabilities.h`** — pure POD data types:
- `Protocol` (Clockless | Spi)
- `FreqRange` / `NanosRange` — inclusive ranges with explicit `any()` / `exactly()` / `atLeast()` / `atMost()` sentinels
- `PinSet` with `PinSetKind::{None, AnyOutputPin, Allowlist}` discriminator, bitset-backed via `fl::bitset_fixed<128>` (covers every MCU FastLED targets)
- `PinGroup` — one independent hardware resource per driver: pin allowlists for data + clock, per-group frequency, max_concurrent, shared_clock_domain_id, shared_budget_id, pins_must_be_contiguous
- `ClocklessTimingCapability` — driver's clock-tree model (base_clock + divider range + counter width)
- `DriverCapabilities` — tiny driver-wide header (protocol gates, freq ranges, timing model, flags)
- `ChannelRequest` — passive input. Single-pin uses `ChannelRequest::singlePin()` factory; multi-pin (ganged PARLIO/FlexIO-style) sets bits in the `data_pins` bitset directly. **One type for both single-pin and bulk requests.**
- `HandleResult` — Yes / NoProtocol / NoPin / NoPinPair / NoFrequency / NoTiming / NoCapacity

**New free function** `src/fl/channels/can_match.h` + `can_match.cpp.hpp`:
- `fl::canMatch(caps, groups, request)` — operates purely on data, no driver instance required
- Walks groups in cheapest-first order (protocol → pins → contiguity → frequency → timing), returns first Yes or strongest rejection
- Clockless timing check uses a divider-loop against chipset tolerance windows — for each divider in `[min, max]`, can `N * tick_ns` land inside each phase's window?

**`IChannelDriver` extension** (`src/fl/channels/driver.h`):
- Adds non-pure virtuals: `getDriverCapabilities()`, `getPinGroups()`, `canMatch()`
- Default impls preserve backward compat (synthesize from legacy `Capabilities`, return empty pin groups, delegate canMatch to free function)
- No existing driver code changes

## Tests

`tests/fl/channels/can_match.cpp` — no driver class instantiated. All tests construct synthetic `DriverCapabilities` + `PinGroup[]` fixtures and call the free function.

Coverage includes single-pin matching, multi-pin (bulk) matching with bitset, FreqRange semantics, clockless timing divider-loop (RMT 80 MHz hits WS2812; 1 MHz tick fails 60-100 ns T0H), PinSet semantics, contiguity, and multi-group "first match wins / strongest rejection surfaces" behavior.

## Out of scope (follow-up PRs)

- Per-driver migration to real static-constexpr descriptors (RMT5 first)
- Manager-side resolver (`resolveDriverAssignments`) with greedy + one-hop swap
- Drop legacy `IChannelDriver::canHandle()` once all drivers migrated

## Test plan

- [x] `bash test can_match` — 25 new test cases pass
- [x] `bash test --cpp` — 329/329 full C++ suite passes
- [x] `bash lint --cpp` — clean

Refs #3186.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a capability-driven channel compatibility check for protocols, pin-group constraints, frequency windows, and clockless timing, returning standardized match/rejection outcomes.
  * Extended the driver interface with capability/pin-group accessors and a unified `canMatch` implementation based on the compatibility rules.

* **Tests**
  * Added unit coverage for protocol gating, empty/invalid pin requests, SPI fixed vs shared-clock matching, frequency-range handling, clockless timing compatibility (including boundary cases), pin-set contiguity rules, capacity limits, and cross-group rejection ranking.

### CODE REVIEW

File: src/fl/channels/can_match.cpp.hpp [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+            continue;
+        }
+        fl::u32 tick_ns = 1000000000u / effective_clock;
+        if (tick_ns == 0u) {
+            continue;
+        }
+        if (canHitPhase(tick_ns, cap.max_phase_ticks, chip.t0h_window) &&
+            canHitPhase(tick_ns, cap.max_phase_ticks, chip.t0l_window) &&
+            canHitPhase(tick_ns, cap.max_phase_ticks, chip.t1h_window) &&
+            canHitPhase(tick_ns, cap.max_phase_ticks, chip.t1l_window)) {
```
[the remark this answers is not in the document]
Comment (zackees) [reply src/fl/channels/can_match.cpp.hpp]: Addressed in 85da3e861 — refactored  to use u64 ratio math ( ceil/floor) instead of the truncated tick_ns intermediate. New test case 'precise tick math (u64 ratio) — exactly at upper bound passes' covers the boundary behavior.

File: src/fl/channels/can_match.cpp.hpp [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        HandleResult r = can_match_detail::checkOneGroup(caps, groups[i], request);
+        if (r == HandleResult::Yes) {
+            return r;
+        }
+        if (can_match_detail::strength(r) > can_match_detail::strength(best)) {
+            best = r;
+        }
+    }
+    return best;
+}
```
[the remark this answers is not in the document]
Comment (zackees) [reply src/fl/channels/can_match.cpp.hpp:230]: Addressed in 85da3e861 — added the driver-wide  gate at the top of  (before the group walk). New test case 'driver-wide max_total_channels is enforced before group walk' confirms a 5-pin bulk is rejected when caps cap is 4 even if a group offers 16 slots.

File: src/fl/channels/capabilities.h [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+    NoPinPair    = 3,   ///< clock pin not in the matching group's allowlist
+    NoFrequency  = 4,   ///< requested freq outside driver/group range
+    NoTiming     = 5,   ///< chipset tolerance windows can't be hit
+    NoCapacity   = 6,   ///< group full / driver-total cap reached
+};
+
+/// Convenience: most call sites really want a bool.
+constexpr bool isYes(HandleResult r) FL_NOEXCEPT {
+    return r == HandleResult::Yes;
+}
```
[the remark this answers is not in the document]
Comment (zackees) [reply src/fl/channels/capabilities.h:368]: Addressed in 001d78354 — moved , , , , and  out to a new  companion (wired into ). Kept in the header: constexpr single-return factories (, , , , , ), the  template (must be in header), and member-init-only constructors (, , ). All 329 C++ tests pass.

File: src/fl/channels/capabilities.cpp.hpp [hunk 4]
```diff
@@ -0,0 +1,93 @@
+/// @file capabilities.cpp.hpp
+/// @brief Out-of-line definitions for capabilities.h types (#3186).
+///
+/// Keeps `src/fl/channels/capabilities.h` declaration-only for the
+/// non-templated, non-trivial functions. Constexpr factories and
+/// templates stay in the header where the language requires them.
+
+// IWYU pragma: private
+
+#pragma once
```
[the remark this answers is not in the document]
Comment (zackees) [reply src/fl/channels/capabilities.cpp.hpp]: Addressed — dropped the raw pragma once from both capabilities.cpp.hpp and can_match.cpp.hpp to match the convention used by channel.cpp.hpp, data.cpp.hpp, and the rest of the channels subsystem (.cpp.hpp routers rely on being included exactly once from _build.cpp.hpp).
